examples/hals/hal_cpu_client/source/hal_cpu_executable.cpp

```diff
@@ -0,0 +1,52 @@
+// Copyright (C) Codeplay Software Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License") with LLVM
+// Exceptions; you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://github.com/codeplaysoftware/oneapi-construction-kit/blob/main/LICENSE.txt
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations
+// under the License.
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <hal_remote/hal_server.h>
+#include <hal_remote/hal_socket_client.h>
+#include <hal_remote/hal_socket_transmitter.h>
+#include <unistd.h>
+
+int main(int argc, char **argv) {
+  uint32_t api_version;
+  auto *hal = get_hal(api_version);
+  uint16_t port = 0;
+  if (const char *env = getenv("HAL_REMOTE_PORT")) {
+    port = atoi(env);
+  }
+
+  hal::hal_socket_transmitter transmitter(port);
+  hal::hal_server::error_code last_error = hal::hal_server::status_success;
+  do {
+    auto res = transmitter.start_server(true);
+    if (res != hal::hal_socket_transmitter::success) {
+      printf("Unable to start server on requested port %d\n", port);
+      fflush(stdout);
+      exit(1);
+    }
+
+    hal::hal_server server(&transmitter, hal);
+    last_error = server.process_commands();
+    // Check if the failure was due the transmitter, if so check if connection
+    // closed, otherwise we have a real error.
+    if (last_error == hal::hal_server::status_transmitter_failed &&
+        transmitter.get_last_error() !=
+            hal::hal_socket_transmitter::connection_closed) {
+      fprintf(stderr, "Error with tcp/ip connection\n");
+      exit(1);
+    }
+
+  } while (last_error == hal::hal_server::status_transmitter_failed);
```

Comment:
Are we sure that after `status_transmitter_failed`, the code is in a state where the server can be meaningfully restarted? I would personally prefer to have a service manager restart the whole process automatically (several service managers exist) but I'm okay with this approach if we know the process is in a good enough state.

Reply:
it's a bit convoluted but I check the connection_closed above so if it fails at this point it's because the connection is closed. However I think we agreed before that this could lead to all sorts of issues with allocs being left over etc, so I'll take out the loop and advise the customer to use a bash while loop as an option.